### ISSUE TITLE
[Docker] node.js cannot access API

### ISSUE THREAD

**Issue by fmakowski** [RyanNoelk/OpenEats#303]
Hey,

after few attempts to get it up and running, I have finally decided to ask for your help concerning OpenEats production setup with Docker. 

I followed [Running the App guide](https://github.com/RyanNoelk/OpenEats/blob/master/docs/Running_the_App.md). I am using docker-compose with following docker-prod.yml file:

```
version: '2'
services:
  api:
    image: openeats/api
    command: /startup/prod-entrypoint.sh
    restart: unless-stopped
    volumes:
      - static-files:/code/static-files
      - site-media:/code/site-media
    depends_on:
      - db
    env_file:
      env_prod.list
  node:
    image: openeats/node
    command: npm start
    restart: unless-stopped
    depends_on:
      - api
    env_file:
      env_prod.list
    
  db:
    image: mariadb
    env_file:
      env_prod.list
  nginx:
    image: openeats/nginx
    command: ./start.sh
    restart: unless-stopped
    ports:
      - "8089:80"
    volumes:
      - static-files:/var/www/html/openeats-static/static-files
      - site-media:/var/www/html/openeats-static/site-media
    depends_on:
      - api
      - node
    env_file:
      env_prod.list
volumes:
  database:
  static-files:
  site-media:
```

and following env_prod.list:

```
# Database config
MYSQL_DATABASE=openeats
MYSQL_ROOT_PASSWORD=ROOT_PASS

# Django config
API_URL=0.0.0.0:8000
API_PORT=8000
DJANGO_SECRET_KEY=SECRET_KEY
DJANGO_SETTINGS_MODULE=base.settings
DJANGO_DEBUG=False
ALLOWED_HOST=localhost
HTTP_X_FORWARDED_PROTO=true

# Node config
NODE_ENV=production
NODE_URL=localhost:8080
NODE_PORT=8080
NODE_API_URL=http://localhost
NODE_LOCALE=en
```
I fire up `docker-compose -f docker-prod.yml up -d` and the containers are properly created. The app is available under http://localhost:8089, but with no data. For that reason, I type `docker exec -ti xyz_api_1 bash` and use `./manage.py migrate` and `./manage.py createsuperuser` to have at least one account available. Unfortunately, when trying to log in, there is no response from the site - not even an error (so it does not resemble #279 in that matter, where there was an error after trying to log in). I tried to populate database with `./manage.py loaddata ....json` using all available fixtures, but still, nothing happened - also, the data loaded was not shown by the frontend (but it might be that way if the data is available only after logging in...). 

I suspect, that for some reason there is no connection from Node.js to the API - I tried to use `expose: "8000"` for `api` container, but it did not change anything. 

I would be grateful for any help regarding that issue, maybe it's just my mistake in setup or changing of docker-prod.yml file.

Best regards!
Filip

**Comment by RyanNoelk** [comment]
Hey there, couple things to try:
- First are there any error message/warnings when you are accessing the API? You can open up the inspect tool in chrome (or any browser) and see the error messages that are happening.
- Access the admin panel: The admin panel is running a the Django admin panel and is powered by the API. Try this link and see what error/message you get. http://localhost:8089/admin/login/?next=/admin/
- Try restating the containers and have the logs print to std out. You can do this by running `docker-compose -f docker-prod.yml up`

Thanks!

**Comment by fmakowski** [comment]
I am indeed able to go to the Django admin panel, and the log when restarting containers does not indicate any errors. 

The nginx HTTP request logs for both api and node are shown properly in the console. The inspector shows the following errors though (sorry for the verbosity level, I have no idea what might be helpful at that point):

```
bundle.js:17 GET http://localhost/api/v1/recipe/mini-browse/?limit=4 net::ERR_CONNECTION_REFUSED
d.end @ bundle.js:17
value @ bundle.js:14
value @ bundle.js:14
commitLifeCycles @ bundle.js:17
n @ bundle.js:17
c @ bundle.js:17
u @ bundle.js:17
m @ bundle.js:17
h @ bundle.js:17
updateContainer @ bundle.js:17
(anonymous) @ bundle.js:17
unbatchedUpdates @ bundle.js:17
ft @ bundle.js:17
render @ bundle.js:17
q @ bundle.js:17
(anonymous) @ bundle.js:17
t @ bundle.js:1
(anonymous) @ bundle.js:1
(anonymous) @ bundle.js:1
bundle.js:14 http://localhost/api/v1/recipe/mini-browse/?limit=4 Error: Request has been terminated
Possible causes: the network is offline, Origin is not allowed by Access-Control-Allow-Origin, the page is being unloaded, etc.
(anonymous) @ bundle.js:14
d.callback @ bundle.js:17
d.crossDomainError @ bundle.js:17
n.onreadystatechange @ bundle.js:17
XMLHttpRequest.send (async)
d.end @ bundle.js:17
value @ bundle.js:14
value @ bundle.js:14
commitLifeCycles @ bundle.js:17
n @ bundle.js:17
c @ bundle.js:17
u @ bundle.js:17
m @ bundle.js:17
h @ bundle.js:17
updateContainer @ bundle.js:17
(anonymous) @ bundle.js:17
unbatchedUpdates @ bundle.js:17
ft @ bundle.js:17
render @ bundle.js:17
q @ bundle.js:17
(anonymous) @ bundle.js:17
t @ bundle.js:1
(anonymous) @ bundle.js:1
(anonymous) @ bundle.js:1
bundle.js:17 GET http://localhost/api/v1/news/entry/ net::ERR_CONNECTION_REFUSED
d.end @ bundle.js:17
value @ bundle.js:17
value @ bundle.js:17
commitLifeCycles @ bundle.js:17
n @ bundle.js:17
c @ bundle.js:17
u @ bundle.js:17
m @ bundle.js:17
h @ bundle.js:17
updateContainer @ bundle.js:17
(anonymous) @ bundle.js:17
unbatchedUpdates @ bundle.js:17
ft @ bundle.js:17
render @ bundle.js:17
q @ bundle.js:17
(anonymous) @ bundle.js:17
t @ bundle.js:1
(anonymous) @ bundle.js:1
(anonymous) @ bundle.js:1
bundle.js:17 http://localhost/api/v1/news/entry/ Error: Request has been terminated
Possible causes: the network is offline, Origin is not allowed by Access-Control-Allow-Origin, the page is being unloaded, etc.
(anonymous) @ bundle.js:17
d.callback @ bundle.js:17
d.crossDomainError @ bundle.js:17
n.onreadystatechange @ bundle.js:17
XMLHttpRequest.send (async)
d.end @ bundle.js:17
value @ bundle.js:17
value @ bundle.js:17
commitLifeCycles @ bundle.js:17
n @ bundle.js:17
c @ bundle.js:17
u @ bundle.js:17
m @ bundle.js:17
h @ bundle.js:17
updateContainer @ bundle.js:17
(anonymous) @ bundle.js:17
unbatchedUpdates @ bundle.js:17
ft @ bundle.js:17
render @ bundle.js:17
q @ bundle.js:17
(anonymous) @ bundle.js:17
t @ bundle.js:1
(anonymous) @ bundle.js:1
(anonymous) @ bundle.js:1
```

**Comment by fmakowski** [comment]
Ahh, I think I get it! It is requesting http://localhost/api, where it should request http://localhost:PORT/api. Am I right? I will experiment with that, I am not sure which one should be specified there (I guess 8000, but let me see)

**Comment by fmakowski** [comment]
Ok, solved! It was not 8000, but 8089, so the working `env_prod.list` file looks like this:

```
# Database config
MYSQL_DATABASE=openeats
MYSQL_ROOT_PASSWORD=ROOT_PASS

# Django config
API_URL=0.0.0.0:8000
API_PORT=8000
DJANGO_SECRET_KEY=SECRET_KEY
DJANGO_SETTINGS_MODULE=base.settings
DJANGO_DEBUG=False
ALLOWED_HOST=localhost
HTTP_X_FORWARDED_PROTO=true

# Node config
NODE_ENV=production
NODE_URL=localhost:8080
NODE_PORT=8080
NODE_API_URL=http://localhost:8089
NODE_LOCALE=en
```

Thank you for your help!

BTW. there is a race condition during the docker container setup - migrations are invoked before the database itself is initialized (although the container itself is already there) - so Django can't connect to the database, interrupting the procedure and never returning to it again. I would propose the solution used in the MonicaHQ, which is a very simple script to wait for the database to be up: https://github.com/monicahq/monica/blob/master/docker/waitfordb.sh

Cheers!

**Comment by RyanNoelk** [comment]
Thanks for going though this! I created another ticket for the race condition (https://github.com/RyanNoelk/OpenEats/issues/304) I'll take a look at it the next chance I get.

I will also make a note in the docs about the `NODE_API_URL` env var and how it needs to be the extra location of the API server.

Thanks!
